website/projects/aws/awssync_refactored.py

```diff
@@ -14,6 +19,86 @@ class AWSSyncRefactored:
     def __init__(self):
         """Create an AWSSync instance."""
         self.api_talker = AWSAPITalker()
+        self.logger = logging.getLogger("django.aws")
+        self.logger.setLevel(logging.DEBUG)
+        self.org_info = None
+        self.fail = False
+
+    def get_syncdata_from_giphouse(self) -> list[SyncData]:
+        """
+        Create a list of SyncData struct containing email, slug and semester.
+
+        Slug and semester combined are together an uniqueness constraint.
+
+        :return: list of SyncData structs with email, slug and semester
+        """
+        sync_data_list = []
+        current_semester = Semester.objects.get_or_create_current_semester()
+
+        for project in Project.objects.filter(mailinglist__isnull=False, semester=current_semester).values(
+            "slug", "semester", "mailinglist"
+        ):
+            project_slug = project["slug"]
+            project_semester = str(Semester.objects.get(pk=project["semester"]))
+            project_email = MailingList.objects.get(pk=project["mailinglist"]).email_address
+
+            sync_data = SyncData(project_email, project_slug, project_semester)
+            sync_data_list.append(sync_data)
+        return sync_data_list
+
+    def generate_aws_sync_list(self, giphouse_data: list[SyncData], aws_data: list[SyncData]) -> list[SyncData]:
+        """
+        Generate the list of users that are registered on the GiPhouse website, but are not yet invited for AWS.
+
+        This includes their ID and email address, to be able to put users in the correct AWS organization later.
+        """
+        return [x for x in giphouse_data if x not in aws_data]
+
+    def extract_aws_setup(self, parent_ou_id: str) -> AWSTree:
+        """
+        Give a list of all the children of the parent OU.
+
+        :param parent_ou_id: The ID of the parent OU.
+        :return: A AWSTree object containing all the children of the parent OU.
+        """
+        aws_tree = AWSTree(
+            "root",
+            parent_ou_id,
+            [
+                Iteration(
+                    ou["Name"],
+                    ou["Id"],
+                    member_accounts := [
+                        SyncData(
+                            account["Email"],
+                            next(
```

Comment:
I dislike this code repetition, but cant find an easy way to solve it. I would be fine if you keep this for now, but maybe we can look at it next sprint.

Reply:
I couldn't come up with a neat list comprehension solution either to solve this. I proposed the problem to ChatGPT-4 and reached the following after a few interactions, which is 😍. Notice how `tags` is iterating over a list with one element.

```Python
    def get_tag_value(self, tags, key):
        for tag in tags:
            if tag['Key'] == key:
                return tag['Value']
        return None

    def extract_aws_setup(self, parent_ou_id: str) -> AWSTree:
        ...
        aws_tree = AWSTree(
            "root",
            parent_ou_id,
            [
                Iteration(
                    ou["Name"],
                    ou["Id"],
                    member_accounts := [
                        SyncData(
                            account["Email"],
                            self.get_tag_value(tags, "project_slug"),
                            self.get_tag_value(tags, "project_semester"),
                        )
                        for account in self.api_talker.list_accounts_for_parent(parent_id=ou["Id"])
                        for tags in [self.api_talker.list_tags_for_resource(resource_id=account["Id"])]
                    ]
                )
                for ou in self.api_talker.list_organizational_units_for_parent(parent_id=parent_ou_id)
            ]
        )
        ...
```